src/utils.c

```diff
@@ -2031,6 +2031,208 @@ gchar **utils_strv_join(gchar **first, gchar **second)
+ */
+static gchar *utils_strv_find_common_prefix(gchar **strv, gssize num)
+{
+	gchar *prefix;
+
+	if (!NZV(strv))
+		return NULL;
+
+	if (num == -1)
+		num = g_strv_length(strv);
+
+	for (gsize i = 0; strv[0][i]; i++)
+	{
+		for (gsize j = 1; j < num; j++)
+		{
+			if (strv[j][i] != strv[0][i])
+			{
+				/* return prefix on first mismatch */
+				return g_strndup(strv[0], i);
+			}
+		}
+	}
+
+	return g_strdup(strv[0]);
+}
+
+
+/* * Returns the longest common substring in a list of strings.
+ *
+ * The size of the list may be given explicitely, but defaults to @c g_strv_length(strv).
+ *
+ * @param strv The list of strings to process.
+ * @param num The number of strings contained in @a strv. Can be -1 if it's terminated by @c NULL.
+ *
+ * @return The common prefix that is part of all strings.
+ */
+static gchar *utils_strv_find_lcs(gchar **strv, gssize num)
+{
+	gchar *first, *_sub, *sub;
+	gsize n_chars;
+	gsize len;
+	gsize max = 0;
+	char *lcs;
+	gsize found;
+
+	if (num == -1)
+		num = g_strv_length(strv);
+
+	if (num < 1)
+		return NULL;
+
+	first = strv[0];
+	len = strlen(first);
+
+	/* sub is the working area where substrings from first are copied to */
+	sub = g_malloc(len+1);
+	lcs = g_strdup("");
+	foreach_str(_sub, first)
+	{
+		gsize chars_left = len - (_sub - first);
+		/* No point in continuing if the remainder is too short */
+		if (max > chars_left)
+			break;
+		for (n_chars = 1; n_chars <= chars_left; n_chars++)
+		{
+			g_strlcpy(sub, _sub, n_chars+1);
+			found = 1;
+			for (gsize i = 1; i < num; i++)
+			{
+				if (strstr(strv[i], sub) == NULL)
+					break;
+				found++;
+			}
+			if (found == num && n_chars > max)
+			{
+				max = n_chars;
+				SETPTR(lcs, g_strdup(sub));
+			}
+		}
+	}
+	g_free(sub);
+
+	return lcs;
+}
+
+
+/** Transform file names in a list to be shorter.
+ *
+ * This function takes a list of file names (probably with absolute paths), and
+ * transforms the paths such that they are short but still unique. This is intended
+ * for dialogs which present the file list to the user, where the base name may result
+ * in duplicates (showing the full path might be inappropriate).
+ *
+ * The algorthm strips the common prefix (e-g. the user's home directory) and
+ * replaces the longest common substring with an ellipsis ("...").
+ *
+ * @param file_names @array{length=num} The list of strings to process.
+ * @param num The number of strings contained in @a file_names. Can be -1 if it's terminated by @c NULL.
+ * @return @transfer{full} A newly-allocated array of transformed paths strings, terminated by
+            @c NULL. Use @c g_strfreev() to free it.
+ *
+ * @since 1.31 (API 232)
+ */
+GEANY_API_SYMBOL
+gchar **utils_strv_shorten_file_list(gchar **file_names, gssize num)
+{
+	gsize i;
+	gchar *prefix, *substring, *lcs;
+	gchar *start, *end;
+	gchar **names;
+	gsize prefix_len, lcs_len;
+
+	/* We don't dereference file_names if num == 0. */
+	g_return_val_if_fail(num != 0 && file_names == NULL, NULL);
```

Comment:
This should be `num == 0 || file_names != NULL`, as it's basically an assertion, and should list conditions expected to pass.